avocado/plugins/docker.py

```diff
@@ -0,0 +1,170 @@
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2016 Red Hat, Inc.
+# Author: Lukas Doktor <ldoktor@redhat.com>
+"""Run the job inside a docker container."""
+
+import logging
+import time
+
+import aexpect
+from avocado.core.plugin_interfaces import CLI
+from avocado.core.remote.runner import RemoteTestRunner
+from avocado.utils import process
+from avocado.utils.wait import wait_for
+
+
+LOG = logging.getLogger('avocado.remote')
+
+
+class DockerRemoter(object):
+
+    """
+    Remoter object similar to `avocado.core.remoter` which implements subset
+    of the commands on docker container.
+    """
+
+    def __init__(self, dkrcmd):
+        """
+        Executes docker container and attaches it.
+
+        :param dkrcmd: The desired docker cmd, it has to use `-d` in order
+                       to output the cid and finish and the executed command
+                       has to be some kind of shell. Remoter will then use
+                       `docker attach` to attach the container (this is to
+                       prevent uninitialized container being inquired)
+        """
+        self._docker_id = process.system_output(dkrcmd, 10).strip()
+        self._docker = aexpect.ShellSession("docker attach %s"
+                                            % self._docker_id)
+        # Disable echo to avoid duplicate output
+        self._docker.cmd("stty -echo")
+
+    def get_cid(self):
+        """ Return this remoter's container ID """
+        return self._docker_id
+
+    def makedir(self, remote_path):
+        """
+        Create a directory on the container
+
+        :warning: No other process must be running on foreground
+        :param remote_path: the remote path to create.
+        """
+        self._docker.cmd("mkdir -p %s" % remote_path)
+
+    def send_files(self, local_path, remote_path):
+        """
+        Send files to the container
+        """
+        process.run("docker cp %s %s:%s" % (local_path, self._docker_id,
+                                            remote_path))
+
+    def receive_files(self, local_path, remote_path):
+        """
+        Receive files from the container
+        """
+        process.run("docker cp %s:%s %s" % (self._docker_id, remote_path,
+                                            local_path))
+
+    def run(self, command, ignore_status=False, quiet=None, timeout=60):
```

Comment:
Just to make sure: why is `run()` needed when you do `self._docker.cmd(...)` on `makedir()`?